### ISSUE TITLE
[Feature] Flaky Tests - retry all tests at end of the test run

### ISSUE THREAD

**Issue by mstepin** [microsoft/playwright#23354]
It would be ideal to have a configuration option where flaky tests are retried at the end of the run once all other tests are finished. I see it was discussed [here](https://github.com/microsoft/playwright/issues/14215), but that one was closed and not sure the concept was since revisited.

This becomes a lot more relevant when using worker-scoped fixtures, since the thread gets torn down and all the `auto: true` worker-scoped fixtures are retriggered for the new thread. Since many of these startup actions are expensive both in time and resource, it makes more sense for the failed tests to share new threads at the end of the run.

My personal vision of a solution would be a config setting in `playwright.config` file eg `retry-strategy: 'tailing'`. Doing it this way would also open the opportunity for other configurations such as not starting a new thread on failure, or max number of hard fails, conditional retries etc.

**Comment by Vovencia** [comment]
You can use `postinstall` to replace one line in the file:

You need to replace `this._queue.unshift(result.newJob)` with `this._queue.push(result.newJob)`

Source file: https://github.com/microsoft/playwright/blob/release-1.40/packages/playwright/src/runner/dispatcher.ts#L126
Compiled file (in your project): `/node_modules/@playwright/test/lib/runner/dispatcher.js`

---

Можно с помощью `postinstall` заменить в файле одну строчку:

Нужно заменить `this._queue.unshift(result.newJob)` на `this._queue.push(result.newJob)`

Исходный файл: https://github.com/microsoft/playwright/blob/release-1.40/packages/playwright/src/runner/dispatcher.ts#L126
Скомпилированный файл (в вашем проекте): `/node_modules/@playwright/test/lib/runner/dispatcher.js`

**Comment by zargham-leanix** [comment]
+1

**Comment by joeflan** [comment]
This would be pretty useful for our team as well

**Comment by olekach** [comment]
+1. It's useful for the case when there is an env issue and all subsequent retries will fail but if they are postponed for a bit later, chances are the test will succeed. Would be good to dynamically manage this.

**Comment by AndiBMGitHub** [comment]
+1 I would like this to be implemented :)

**Comment by DerTimonius** [comment]
+1
This would be really useful when resetting the test data only once in the global setup and the flaky test now does not have a clean state it had on the initial run

**Comment by marcusNumminen** [comment]
+1 one for this

**Comment by philipfong** [comment]
+1 I'd like to see this too. This is also useful for when some specific tests may fail when sharded and the system is under heavy load, and then may have a better chance of succeeding when the number of shards tail off.

I know this all speaks to generally poor systems under test, but that's kind of the reality that we have to test with much of the time.

**Comment by TuxThunderbird** [comment]
+1 for this.

**Comment by vbdata** [comment]
+1 I'd like to see this too.

**Comment by lynn57blocks** [comment]
+1 I'd like to see this too.

**Comment by kosteklvp** [comment]
+1 I'd like to see this too.

**Comment by longmaxx** [comment]
+1 Would be really useful as a config param.

**Comment by tragsdale7** [comment]
+1 would like this

**Comment by mastrzyz** [comment]
+1

**Comment by deniss-nikitin** [comment]
+1

**Comment by dgozman** [comment]
We have discussed this feature request with a team. It looks like an option to run retries one by one at the end, without any parallelism, might be beneficial to reduce flakiness. Note that this will not improve the total run time, but rather slow it down in the hope of making the test suite more stable.

Please let us know whether you would use such a feature. If you would like to see something else, please explain your usecase in details. Thank you!

**Comment by matmair** [comment]
i would be very interested in this pattern. We sometimes experience flaky test due to the backend overloading / some things locking up processes. The proposed pattern could adress those

**Comment by marcusNumminen** [comment]
Couldn't there be option to either run the failing test in the end with or without parallelism?

**Comment by dgozman** [comment]
@marcusNumminen If you would like the feature to run retries in the end, but in parallel, please explain your usecase in details. We'd like to understand the problem before shipping a solution.

**Comment by mstepin** [comment]
@dgozman I assume your proposed solution involves tearing down the threads at the end and creating a new pristine single thread to run the tests one by one?

**Comment by dgozman** [comment]
> [@dgozman](https://github.com/dgozman) I assume your proposed solution involves tearing down the threads at the end and creating a new pristine single thread to run the tests one by one?

Yes, we'll create a new worker and start running retries there. If some of them fail again, a new worker will be created as usual.

**Comment by marcusNumminen** [comment]
Hi!
Sorry for the delay. In some cases we are running the test right after a deploy and when the tests start not all the deployed stuff is fully ready and some test fails due to this. But a solution with one worker running the retries in the end of the execution will of solve the problem for us but in this situation running the test in parallel would be faster :)

**Comment by Shaun-Regan** [comment]
+1

**Comment by JBelisle** [comment]
I have the same situation as @marcusNumminen in that, no matter how many readiness checks I add to our pipeline, the new deploy is not always fully ready when the tests start, so forcing the retries to not be parallel wouldn't matter one way or the other.

For this situation though, just having the option to retry after the run completes instead of immediately after the failure will make our pipelines significantly less flakey, and that is more important to me than shaving time by parallelizing the retries.

**Comment by marcusNumminen** [comment]
Any news about this one? Would have been really nice to get this in :)

**Comment by myantahtova** [comment]
+1, would be a great feature

**Comment by Francesco-Borzi** [comment]
This would help as some tests get flaky when there are many workers running. Re-running them at the end, when less tests are running in parallel, would increase the overall stability.

**Comment by alexeyraspopov** [comment]
> Please let us know whether you would use such a feature. If you would like to see something else, please explain your usecase in details. Thank you!

@dgozman is it still something on your roadmap?

My usecase: our UI depends on a lot of backend services and occasionally a test can fail due to poor timing from the side we don't control. When it happens, the test attempts to retry immediately, potentially still executing the same logic during a downtime of that faulty service. If these failed tests would attempt to retry at the end, there's a chance some of those root causes would disappear by then.

**Comment by dgozman** [comment]
For the reference, #41798 updated the behavior.
